capa/features/extractors/binexport2/insn.py

```diff
@@ -313,11 +317,89 @@ def extract_insn_offset_features(
     yield from ()
 
 
+def is_security_cookie(
+    fhi: FunctionContext,
+    bbi: BasicBlockContext,
+    instruction: BinExport2.Instruction,
+) -> bool:
+    """
+    check if an instruction is related to security cookie checks.
+    """
+    be2 = fhi.ctx.be2
+
+    # security cookie check should use SP or BP
+    op1 = be2.operand[instruction.operand_index[1]]
+    op1_exprs = [be2.expression[expr_i] for expr_i in op1.expression_index]
+    if all(
+        expr.type != BinExport2.Expression.Type.REGISTER or
+        expr.symbol.lower() not in ("bp", "esp", "ebp", "rbp", "rsp")
+        for expr in op1_exprs
+    ):
+        return False
+
+    # check_nzxor_security_cookie_delta
+    # if insn falls at the start of first entry block of the parent function.
+    flow_graph = be2.flow_graph[fhi.flow_graph_index]
+    basic_block_index = bbi.basic_block_index
+    bb = be2.basic_block[basic_block_index]
+    if flow_graph.entry_basic_block_index == basic_block_index:
+        first_addr = min((
+            be2.instruction[ir.begin_index].address
+            for ir in bb.instruction_index))
+        if instruction.address < first_addr + SECURITY_COOKIE_BYTES_DELTA:
+            return True
+    # or insn falls at the end before return
+    fbbi = set(flow_graph.basic_block_index)
+    for edge in flow_graph.edge:
+        if (edge.source_basic_block_index == basic_block_index and
+            edge.target_basic_block_index not in fbbi):
+            last_addr = max((
+                be2.instruction[ir.end_index - 1].address
+                for ir in bb.instruction_index))
+            if instruction.address > last_addr - SECURITY_COOKIE_BYTES_DELTA:
+                return True
```

Comment:
This assumes that the last instruction, ordered by virtual address, will be the exit and the security cookie check is found nearby.  While I think this probably mostly works in practice, I don't think its robust. I'd suggest a more complete solution might be: find all terminal basic blocks (no successors) in the function flow graph and consider each of those a potential exit block.

It may also be worth pre-computing the instruction indicies that fall within these ranges, since a function may have hundreds or thousands of instructions, and this routine might get invoked a lot. On the other hand, I think this is only invoked once the mnemonic and operands are checked, so maybe its not a problem in practice.